state/machine.go

```diff
@@ -2203,6 +2182,70 @@ func (m *Machine) RemoveUpgradeSeriesLock() error {
 	return nil
 }
 
+// UpgradeSeriesStatus returns the status of a series upgrade.
+func (m *Machine) UpgradeSeriesStatus(unitName string) (params.UnitSeriesUpgradeStatus, error) {
+	coll, closer := m.st.db().GetCollection(machineUpgradeSeriesLocksC)
+	defer closer()
+
+	var lock upgradeSeriesLock
+	err := coll.FindId(m.Id()).One(&lock)
+	if err == mgo.ErrNotFound {
+		return "", errors.NotFoundf("upgrade series lock for machine %q", m.Id())
+	}
+	if err != nil {
+		return "", errors.Trace(err)
+	}
+
+	for _, unit := range lock.PrepareUnits {
+		if unit.Id == unitName {
+			return unit.Status, nil
+		}
+	}
+
+	return "", errors.NotFoundf("unit %q of machine %q", unitName, m.Id())
+}
+
+// SetUpgradeSeriesStatus sets the status of a series upgrade.
+func (m *Machine) SetUpgradeSeriesStatus(unitName string, status params.UnitSeriesUpgradeStatus) error {
+	buildTxn := func(attempt int) ([]txn.Op, error) {
+		if attempt > 0 {
+			if err := m.Refresh(); err != nil {
+				return nil, errors.Trace(err)
+			}
+		}
+		if err := m.isStillAlive(); err != nil {
+			return nil, errors.Trace(err)
+		}
+		coll, closer := m.st.db().GetCollection(machineUpgradeSeriesLocksC)
```

Comment:
This function should use `UpgradeSeriesStatus` above to short circuit the check.

```go
unitStatus, err := m.UpgradeSeriesStatus(unitName)
if err != nil {
  return nil, errors.Trace(err)
}
if unitStatus == status {
return nil, jujutxn.ErrNoOperations
}
// pass in the current status to assert against, and the status to set it to.
return setUpgradeSeriesTxnOps(m.Id(), unitStatus, status), nil
}
```